Tools/speech_recognizer/GSSpeechRecognizer.m

```diff
@@ -0,0 +1,101 @@
+#import "GSSpeechRecognizer.h"
+
+static GSSpeechRecognitionServer *server;
+static int clients;
+
+@interface GSSpeechRecognizer (Private)
++ (void)connectionDied: (NSNotification*)aNotification;
+@end
+
+@implementation GSSpeechRecognizer
++ (void)initialize
+{
+  server = [GSSpeechRecognitionServer sharedServer];
+
+  RETAIN(server);
+  if (server != nil)
+    {
+      clients++;
+    }
+  
+  [[NSNotificationCenter defaultCenter]
+		addObserver: self
+		   selector: @selector(connectionDied:)
+                       name: NSConnectionDidDieNotification
+                     object: nil];
+}
+
+/**
+ * If the remote end exits before freeing the GSSpeechRecognizer then we need
+ * to send it a -release message to make sure it dies.
+ */
++ (void)connectionDied: (NSNotification*)aNotification
+{
+  NSEnumerator *e = [[[aNotification object] localObjects] objectEnumerator];
+  NSObject *o = nil;
+  for (o = [e nextObject] ; nil != o ; o = [e nextObject])
+    {
+      if ([o isKindOfClass: self])
+        {
+          [o release];
+        }
+    }
+}
+
+/**
+ * If no clients have been active for some time, kill the speech server to
+ * conserve resources.
+ */
++ (void)exitIfUnneeded: (NSTimer*)sender
+{
+  if (clients == 0)
+    {
+      exit(0);
+    }
+}
+
+- (id)init
+{
+  self = [super init];
+  if (self != nil)
+    {
+      NSLog(@"self = %@",self);
+      RETAIN(server);
```

Comment:
This line doesn't make any sense. You should rather increase your clients variable here.

Reply:
I apologize... I am trying to diagnose a crash that has been occurring.

Reply:
Added to count here.